File: frameworks/csharp-dotnet/FraiseQL.Benchmark.Tests/ErrorEdgeCasesTest.cs
using FraiseQL.Benchmark.Models;
using Xunit;
using System;
using System.Linq;

namespace FraiseQL.Benchmark.Tests;

public class ErrorEdgeCasesTest
{
    private readonly TestFactory _factory = new();

    // ============================================================================
    // Error: HTTP Status Codes
    // ============================================================================

    [Fact]
    public void TestHttpStatusCodeSuccess()
    {
        _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        Assert.Equal(1, _factory.GetUserCount());
    }

    [Fact]
    public void TestHttpStatusCodeNotFound()
    {
        var user = _factory.GetUser(Guid.NewGuid());
        Assert.Null(user);
    }

    // ============================================================================
    // Error: 404 Not Found
    // ============================================================================

    [Fact]
    public void TestUserNotFoundReturnsNull()
    {
        var user = _factory.GetUser(Guid.NewGuid());
        Assert.Null(user);
    }

    [Fact]
    public void TestPostNotFoundReturnsNull()
    {
        var post = _factory.GetPost(Guid.NewGuid());
        Assert.Null(post);
    }

    // ============================================================================
    // Error: Invalid Input
    // ============================================================================

    [Fact]
    public void TestInvalidLimitNegative()
    {
        int limit = -5;
        int clamped = Math.Max(0, Math.Min(100, limit));
        Assert.Equal(0, clamped);
    }

    [Fact]
    public void TestInvalidLimitZero()
    {
        int limit = 0;
        int clamped = Math.Max(0, Math.Min(100, limit));
        Assert.Equal(0, clamped);
    }

    [Fact]
    public void TestVeryLargeLimit()
    {
        int limit = 999999;
        int clamped = Math.Max(0, Math.Min(100, limit));
        Assert.Equal(100, clamped);
    }

    // ============================================================================
    // Edge Case: UUID Validation
    // ============================================================================

    [Fact]
    public void TestAllUserIdsAreUUID()
    {
        _factory.CreateTestUser("user0", "user0@example.com", "User", "");
        _factory.CreateTestUser("user1", "user1@example.com", "User", "");
        _factory.CreateTestUser("user2", "user2@example.com", "User", "");

        var users = _factory.GetAllUsers().ToList();
        Assert.All(users, u => Assert.NotEqual(Guid.Empty, u.Id));
    }

    [Fact]
    public void TestAllPostIdsAreUUID()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");

        _factory.CreateTestPost(author.Id, "Post0", "Content");
        _factory.CreateTestPost(author.Id, "Post1", "Content");
        _factory.CreateTestPost(author.Id, "Post2", "Content");

        var posts = _factory.GetAllPosts().ToList();
        Assert.All(posts, p => Assert.NotEqual(Guid.Empty, p.Id));
    }

    // ============================================================================
    // Edge Case: Special Characters
    // ============================================================================

    [Fact]
    public void TestSpecialCharSingleQuotes()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "I'm a developer");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharDoubleQuotes()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "He said \"hello\"");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharHtmlTags()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Check <this> out");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharAmpersand()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "Tom & Jerry");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharEmoji()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "🎉 Celebration! 🚀 Rocket");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharAccents()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Àlice Müller", "");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    [Fact]
    public void TestSpecialCharDiacritics()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "José García", "");
        Assert.NotNull(_factory.GetUser(user.Id));
    }

    // ============================================================================
    // Edge Case: Boundary Conditions
    // ============================================================================

    [Fact]
    public void TestBoundaryVeryLongBio5000Chars()
    {
        var longBio = GenerateLongString(5000);
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", longBio);
        var retrieved = _factory.GetUser(user.Id);
        Assert.NotNull(retrieved?.Bio);
        Assert.Equal(5000, retrieved.Bio.Length);
    }

    [Fact]
    public void TestBoundaryVeryLongUsername255Chars()
    {
        var longName = GenerateLongString(255);
        var user = _factory.CreateTestUser(longName, "user@example.com", "User", "");
        var retrieved = _factory.GetUser(user.Id);
        Assert.Equal(255, retrieved?.Username.Length);
    }

    [Fact]
    public void TestBoundaryVeryLongPostTitle()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var longTitle = GenerateLongString(500);
        var post = _factory.CreateTestPost(author.Id, longTitle, "Content");
        var retrieved = _factory.GetPost(post.Id);
        Assert.Equal(500, retrieved?.Title.Length);
    }

    [Fact]
    public void TestBoundaryVeryLongContent5000Chars()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var longContent = GenerateLongString(5000);
        var post = _factory.CreateTestPost(author.Id, "Title", longContent);
        var retrieved = _factory.GetPost(post.Id);
        Assert.Equal(5000, retrieved?.Content.Length);
    }

    // ============================================================================
    // Edge Case: Null/Empty Fields
    // ============================================================================

    [Fact]
    public void TestNullBioIsHandled()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var retrieved = _factory.GetUser(user.Id);
        Assert.Null(retrieved?.Bio);
    }

    [Fact]
    public void TestEmptyStringBio()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var retrieved = _factory.GetUser(user.Id);
        Assert.Null(retrieved?.Bio);
    }

    [Fact]
    public void TestPresentBio()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "My bio");
        var retrieved = _factory.GetUser(user.Id);
        Assert.NotNull(retrieved?.Bio);
        Assert.Equal("My bio", retrieved.Bio);
    }

    // ============================================================================
    // Edge Case: Relationship Validation
    // ============================================================================

    [Fact]
    public void TestPostAuthorIdIsValidUUID()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Post", "Content");
        Assert.NotNull(post.Author);
        Assert.NotEqual(Guid.Empty, post.Author.Id);
    }

    [Fact]
    public void TestPostReferencesCorrectAuthor()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Post", "Content");
        Assert.Equal(post.FkAuthor, author.PkUser);
    }

    [Fact]
    public void TestMultiplePostsReferenceDifferentAuthors()
    {
        var author1 = _factory.CreateTestUser("author1", "author1@example.com", "Author1", "");
        var author2 = _factory.CreateTestUser("author2", "author2@example.com", "Author2", "");

        var post1 = _factory.CreateTestPost(author1.Id, "Post1", "Content");
        var post2 = _factory.CreateTestPost(author2.Id, "Post2", "Content");

        Assert.NotEqual(post1.FkAuthor, post2.FkAuthor);
        Assert.Equal(post1.FkAuthor, author1.PkUser);
        Assert.Equal(post2.FkAuthor, author2.PkUser);
    }

    // ============================================================================
    // Edge Case: Data Type Validation
    // ============================================================================

    [Fact]
    public void TestUsernameIsString()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var retrieved = _factory.GetUser(user.Id);
        Assert.Equal("alice", retrieved?.Username);
    }

    [Fact]
    public void TestPostTitleIsString()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Test Post", "Content");
        var retrieved = _factory.GetPost(post.Id);
        Assert.Equal("Test Post", retrieved?.Title);
    }

    // ============================================================================
    // Edge Case: Uniqueness
    // ============================================================================

    [Fact]
    public void TestMultipleUsersHaveUniqueIds()
    {
        var user1 = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var user2 = _factory.CreateTestUser("bob", "bob@example.com", "Bob", "");
        var user3 = _factory.CreateTestUser("charlie", "charlie@example.com", "Charlie", "");

        Assert.NotEqual(user1.Id, user2.Id);
        Assert.NotEqual(user2.Id, user3.Id);
        Assert.NotEqual(user1.Id, user3.Id);
    }

    [Fact]
    public void TestMultiplePostsHaveUniqueIds()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post1 = _factory.CreateTestPost(author.Id, "Post1", "Content1");
        var post2 = _factory.CreateTestPost(author.Id, "Post2", "Content2");
        var post3 = _factory.CreateTestPost(author.Id, "Post3", "Content3");

        Assert.NotEqual(post1.Id, post2.Id);
        Assert.NotEqual(post2.Id, post3.Id);
        Assert.NotEqual(post1.Id, post3.Id);
    }

    // ============================================================================
    // Helper Methods
    // ============================================================================

    private string GenerateLongString(int length)
    {
        var chars = new char[length];
        for (int i = 0; i < length; i++)
        {
            chars[i] = (char)('0' + (i % 10));
        }
        return new string(chars);
    }
}

File: frameworks/actix-web-rest/tests/error_edge_cases_test.rs
mod test_helpers;
use test_helpers::*;

// ============================================================================
// Error: HTTP Status Codes
// ============================================================================

#[test]
fn test_http_status_code_success() {
    let factory = TestFactory::new();
    factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert_eq!(factory.user_count(), 1);
}

#[test]
fn test_http_status_code_not_found() {
    let factory = TestFactory::new();
    let user = factory.get_user("nonexistent-id");
    assert!(user.is_none());
}

// ============================================================================
// Error: 404 Not Found
// ============================================================================

#[test]
fn test_user_not_found_returns_none() {
    let factory = TestFactory::new();
    let user = factory.get_user("nonexistent-user-id");
    assert!(user.is_none());
}

#[test]
fn test_post_not_found_returns_none() {
    let factory = TestFactory::new();
    let post = factory.get_post("nonexistent-post-id");
    assert!(post.is_none());
}

// ============================================================================
// Error: Invalid Input
// ============================================================================

#[test]
fn test_invalid_limit_negative() {
    let limit = -5;
    let clamped = if limit < 0 { 0 } else { limit };
    assert_eq!(clamped, 0);
}

#[test]
fn test_invalid_limit_zero() {
    let limit = 0;
    let clamped = if limit < 0 { 0 } else { limit };
    assert_eq!(clamped, 0);
}

#[test]
fn test_very_large_limit() {
    let limit = 999999;
    let clamped = if limit > 100 { 100 } else { limit };
    assert_eq!(clamped, 100);
}

// ============================================================================
// Edge Case: UUID Validation
// ============================================================================

#[test]
fn test_all_user_ids_are_uuid() {
    let factory = TestFactory::new();
    factory.create_test_user("user0", "user0@example.com", "User", "");
    factory.create_test_user("user1", "user1@example.com", "User", "");
    factory.create_test_user("user2", "user2@example.com", "User", "");

    for user in factory.get_all_users() {
        assert!(ValidationHelper::assert_uuid(&user.id).is_ok());
    }
}

#[test]
fn test_all_post_ids_are_uuid() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");

    factory.create_test_post(&author.id, "Post0", "Content");
    factory.create_test_post(&author.id, "Post1", "Content");
    factory.create_test_post(&author.id, "Post2", "Content");

    for post in factory.get_all_posts() {
        assert!(ValidationHelper::assert_uuid(&post.id).is_ok());
    }
}

// ============================================================================
// Edge Case: Special Characters
// ============================================================================

#[test]
fn test_special_char_single_quotes() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "I'm a developer");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_double_quotes() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "He said \"hello\"");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_html_tags() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Check <this> out");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_ampersand() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Tom & Jerry");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_emoji() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "🎉 Celebration! 🚀 Rocket");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_accents() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Àlice Müller", "");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_special_char_diacritics() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "José García", "");
    assert!(factory.get_user(&user.id).is_some());
}

// ============================================================================
// Edge Case: Boundary Conditions
// ============================================================================

#[test]
fn test_boundary_very_long_bio_5000_chars() {
    let factory = TestFactory::new();
    let long_bio = DataGenerator::generate_long_string(5000);
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", &long_bio);
    let retrieved = factory.get_user(&user.id).unwrap();
    assert_eq!(retrieved.bio.as_ref().unwrap().len(), 5000);
}

#[test]
fn test_boundary_very_long_username_255_chars() {
    let factory = TestFactory::new();
    let long_name = DataGenerator::generate_long_string(255);
    let user = factory.create_test_user(&long_name, "user@example.com", "User", "");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert_eq!(retrieved.username.len(), 255);
}

#[test]
fn test_boundary_very_long_post_title() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let long_title = DataGenerator::generate_long_string(500);
    let post = factory.create_test_post(&author.id, &long_title, "Content");
    let retrieved = factory.get_post(&post.id).unwrap();
    assert_eq!(retrieved.title.len(), 500);
}

#[test]
fn test_boundary_very_long_content_5000_chars() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let long_content = DataGenerator::generate_long_string(5000);
    let post = factory.create_test_post(&author.id, "Title", &long_content);
    let retrieved = factory.get_post(&post.id).unwrap();
    assert_eq!(retrieved.content.as_ref().unwrap().len(), 5000);
}

// ============================================================================
// Edge Case: Null/Empty Fields
// ============================================================================

#[test]
fn test_null_bio_is_handled() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert!(retrieved.bio.is_none());
}

#[test]
fn test_empty_string_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert!(retrieved.bio.is_none());
}

#[test]
fn test_present_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "My bio");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert!(retrieved.bio.is_some());
    assert_eq!(retrieved.bio.as_ref().unwrap(), "My bio");
}

// ============================================================================
// Edge Case: Relationship Validation
// ============================================================================

#[test]
fn test_post_author_id_is_valid_uuid() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");
    assert!(ValidationHelper::assert_uuid(&post.author_id).is_ok());
}

#[test]
fn test_post_references_correct_author() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");
    assert_eq!(post.author_id, author.id);
}

#[test]
fn test_multiple_posts_reference_different_authors() {
    let factory = TestFactory::new();
    let author1 = factory.create_test_user("author1", "author1@example.com", "Author1", "");
    let author2 = factory.create_test_user("author2", "author2@example.com", "Author2", "");

    let post1 = factory.create_test_post(&author1.id, "Post1", "Content");
    let post2 = factory.create_test_post(&author2.id, "Post2", "Content");

    assert_ne!(post1.author_id, post2.author_id);
    assert_eq!(post1.author_id, author1.id);
    assert_eq!(post2.author_id, author2.id);
}

// ============================================================================
// Edge Case: Data Type Validation
// ============================================================================

#[test]
fn test_username_is_string() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert_eq!(retrieved.username, "alice");
    assert!(retrieved.username.is_ascii());
}

#[test]
fn test_post_title_is_string() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Test Post", "Content");
    let retrieved = factory.get_post(&post.id).unwrap();
    assert_eq!(retrieved.title, "Test Post");
}

// ============================================================================
// Edge Case: Uniqueness
// ============================================================================

#[test]
fn test_multiple_users_have_unique_ids() {
    let factory = TestFactory::new();
    let user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "");
    let user3 = factory.create_test_user("charlie", "charlie@example.com", "Charlie", "");

    assert_ne!(user1.id, user2.id);
    assert_ne!(user2.id, user3.id);
    assert_ne!(user1.id, user3.id);
}

#[test]
fn test_multiple_posts_have_unique_ids() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post1 = factory.create_test_post(&author.id, "Post1", "Content1");
    let post2 = factory.create_test_post(&author.id, "Post2", "Content2");
    let post3 = factory.create_test_post(&author.id, "Post3", "Content3");

    assert_ne!(post1.id, post2.id);
    assert_ne!(post2.id, post3.id);
    assert_ne!(post1.id, post3.id);
}
